From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: aswin-mcw <aswin.kalies@multicorewareinc.com>
Date: Wed, 9 Sep 2026 13:14:01 +0530
Subject: [PATCH 027/109] Cygwin: gendef: stub implementations of
 routines for AArch64

(cherry picked from commit bb626f0798aeec1bb63c7c4df64e5e256abe5cf6)
---
 winsup/cygwin/scripts/gendef | 42 +++++++++++++++++++++++++++++++++++-
 1 file changed, 41 insertions(+), 1 deletion(-)

diff --git a/winsup/cygwin/scripts/gendef b/winsup/cygwin/scripts/gendef
index 6328fe2fb..2dc9bce53 100755
--- a/winsup/cygwin/scripts/gendef
+++ b/winsup/cygwin/scripts/gendef
@@ -21,6 +21,7 @@ if (!defined($cpu) || !defined($output_def)) {
     die "$0: missing required option\n";
 }
 
+my $is_aarch64 = $cpu eq 'aarch64';
 my $is_x86_64 = $cpu eq 'x86_64';
 # FIXME? Do other (non-32 bit) arches on Windows still use symbol prefixes?
 my $sym_prefix = '';
@@ -89,7 +90,7 @@ sub fefunc {
     my $func = $sym_prefix . shift;
     my $fe = $sym_prefix . shift;
     my $sigfe_func;
-    if ($is_x86_64) {
+    if ($is_x86_64 || $is_aarch64) {
 	$sigfe_func = ($fe =~ /^(.*)_${func}$/)[0];
     }
     my $extra;
@@ -109,6 +110,15 @@ $fe:
 
 EOF
     }
+    # TODO: This is only a stub, it needs to be implemented properly for AArch64.
+    if ($is_aarch64) {
+	$res = <<EOF;
+	.extern $func
+	.global $fe
+$fe:
+EOF
+    }
+
     if (!$main::first++) {
 	if ($is_x86_64) {
 	  $res = <<EOF . longjmp () . $res;
@@ -340,6 +350,23 @@ stabilize_sig_stack:
 	popq	%r12
 	ret
 	.seh_endproc
+EOF
+	}
+	# TODO: These are only stubs, they need to be implemented properly for AArch64.
+	if ($is_aarch64) {
+	  $res = <<EOF . longjmp () . $res;
+	.include "tlsoffsets"
+	.text
+
+_sigfe_maybe:
+	.global _sigbe
+_sigfe:
+_sigbe:
+	.global	sigdelayed
+sigdelayed:
+_sigdelayed_end:
+	.global _sigdelayed_end
+stabilize_sig_stack:
 EOF
 	}
     }
@@ -476,6 +503,19 @@ longjmp:
 	incl	%eax
 0:	ret
 	.seh_endproc
+EOF
+    }
+    if ($is_aarch64) {
+	# TODO: These are only stubs, they need to be implemented properly for AArch64.
+	return <<EOF;
+	.globl	sigsetjmp
+sigsetjmp:
+	.globl  setjmp
+setjmp:
+	.globl	siglongjmp
+siglongjmp:
+	.globl  longjmp
+longjmp:
 EOF
     }
 }
